src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -636,11 +658,39 @@ public void OnReset(Exception resetException, bool canRetry = false)
                         canRetry = false;
                     }
 
-                    _resetException = resetException;
-                    _canRetry = canRetry;
+                    // Per section 8.1 in the RFC:
+                    // If the server has completed the response body (i.e. we've received EndStream)
+                    // but the request body is still sending, and we then receive a RST_STREAM with errorCode = NO_ERROR,
+                    // we treat this specially and simply cancel sending the request body, rather than treating
+                    // the entire request as failed.
+                    if (resetStreamErrorCode == Http2ProtocolErrorCode.NoError && 
+                        _responseCompletionState == StreamCompletionState.Completed)
+                    {
+                        if (_requestCompletionState == StreamCompletionState.InProgress)
+                        {
+                            _requestBodyAbandoned = true;
+                            requestBodyCancellationSource = _requestBodyCancellationSource;
+                            Debug.Assert(requestBodyCancellationSource != null);
+                        }
```

Comment:
Can you help me understand why this isn't just:
```C#
_requestBodyAbandoned = _requestCompletionState == StreamCompletionState.InProgress;
```
and then down below it goes back to just being:
```C#
Cancel();
```
rather than having a special path for requestBodyCancellationSource != null?

I guess I'm not clear on what the difference is between calling requestBodyCancellationSource.Cancel() is and just calling Cancel()?

Reply:
Because Cancel() is going to cancel both the request body and the response body. We only want the former here.

Cancelling the response body is going to cause any subsequent access to the response body to fail. We don't want that.

At one point I had separate CancelRequestBody and CancelResponseBody routines, but I merged them to try to keep things simple, and because when both occur (e.g. typical RST_STREAM) we want the cancellation of both to be atomic. Currently, in all the cases where we call Cancel, we either want to cancel both, or we know that one is already failed and we really just want to cancel the other, in which case calling Cancel is still fine since it won't do anything for the already-failed case. 

But here, we really just want to stop the request body from sending, without cancelling the entire request and response, which is what makes the handling here special and somewhat weird.

Reply:
> Cancelling the response body is going to cause any subsequent access to the response body to fail.

Ah, even though it's already completed?

Reply:
Yes. In general, if we receive the full response body (meaning we get EndStream) and then the request body is failed (cancellation, exception from request body serialization, receiving RST_STREAM, connection failure, etc), we also cancel the response body and discard any remaining buffered data. This mimics, for example, handling TCP RST in HTTP/1.1.

Reply:
Unfortunately, there's another issue here. We are calling Cancel in the exception path in SendRequestBodyAsync anyway, so we're going to end up cancelling the response body regardless. The test I wrote reads the response body before propagating the cancellation through, which is why it didn't fail. I need to think about this further.